control-plane-operator/controllers/awsprivatelink/awsprivatelink_controller.go

```diff
@@ -0,0 +1,122 @@
+package awsprivatelink
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/go-logr/logr"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/informers"
+	kubeclient "k8s.io/client-go/kubernetes"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	hyperv1 "github.com/openshift/hypershift/api/v1alpha1"
+	"github.com/openshift/hypershift/support/upsert"
+)
+
+const (
+	defaultResync = 10 * time.Hour
+)
+
+type PrivateServiceObserver struct {
+	kubeClient *kubeclient.Clientset
+	mgrClient  client.Client
+	log        logr.Logger
+
+	ServiceNamespace string
+	ServiceName      string
+	HCPNamespace     string
+	upsert.CreateOrUpdateProvider
+}
+
+func nameMapper(names []string) handler.MapFunc {
+	nameSet := sets.NewString(names...)
+	return func(obj client.Object) []reconcile.Request {
+		if !nameSet.Has(obj.GetName()) {
+			return nil
+		}
+		return []reconcile.Request{
+			{
+				NamespacedName: types.NamespacedName{
+					Namespace: obj.GetNamespace(),
+					Name:      obj.GetName(),
+				},
+			},
+		}
+	}
+}
+
+func namedResourceHandler(names ...string) handler.EventHandler {
+	return handler.EnqueueRequestsFromMapFunc(nameMapper(names))
+}
+
+func ControllerName(name string) string {
+	return fmt.Sprintf("%s-observer", name)
```

Comment:
This can not have dashes because it ends up being used in a prometheus metric and they don't allow dashes

Reply:
interesting gotcha